File: APOLLO/DNADX/Reproduce/signature_score_and_segment_score_calculation.py
import pandas as pd
import numpy as np
from helper import (calc_signatures, assign_diff_score_dwd, GHI_RS,
                   calc_segments)
import os

def convert_symbol_to_entrez(symbols, id_mapping_file):
    """Convert gene symbols to Entrez IDs using the mapping file"""
    mapping = pd.read_csv(id_mapping_file, sep='\t', header=None, names=['EntrezID', 'Symbol'])
    # Create symbol to EntrezID mapping
    symbol_to_entrez = dict(zip(mapping['Symbol'], mapping['EntrezID'].astype(str)))
    # Convert symbols to EntrezIDs
    return [symbol_to_entrez.get(symbol, symbol) for symbol in symbols]

def convert_entrez_to_symbol(entrez_ids, id_mapping_file):
    """Convert Entrez IDs to gene symbols using the mapping file"""
    mapping = pd.read_csv(id_mapping_file, sep='\t', header=None, names=['EntrezID', 'Symbol'])
    # Create EntrezID to symbol mapping
    entrez_to_symbol = dict(zip(mapping['EntrezID'].astype(str), mapping['Symbol']))
    # Convert EntrezIDs to symbols
    return [entrez_to_symbol.get(entrez, entrez) for entrez in entrez_ids]

def main():
    # Define base paths
    base_path = "/ix1/alee/LO_LAB/Personal/Daisong/Projects/APOLLO/DNADX/Reproduce"
    data_path = os.path.join(base_path, "DNA-based-predictors-of-non-genetic-cancer-phenotypes/data")
    
    # Load input data
    print("Loading input data...")
    edata = pd.read_csv(os.path.join(base_path, "HiSeqV2"), sep='\t', index_col=0)
    
    # Convert gene symbols to Entrez IDs for expression data
    print("Converting gene symbols to Entrez IDs for expression data...")
    edata.index = convert_symbol_to_entrez(edata.index, os.path.join(base_path, "gene_id_mapping.txt"))
    
    # Load and process CNA data
    print("\nLoading CNA data...")
    cna_data = pd.read_csv(os.path.join(base_path, "Gistic2_CopyNumber_Gistic2_all_data_by_genes.txt"), sep='\t')
    
    # Convert gene symbols to Entrez IDs for CNA data
    print("Converting gene symbols to Entrez IDs for CNA data...")
    cna_data['Gene Symbol'] = convert_symbol_to_entrez(cna_data['Gene Symbol'], 
                                                      os.path.join(base_path, "gene_id_mapping.txt"))
    
    # Set gene symbols as index
    cna_data.set_index('Gene Symbol', inplace=True)
    
    # Print first 10 row labels
    print("\nFirst 10 row labels in edata:")
    print(edata.index[:10].tolist())
    print("\nFirst 10 row labels in CNA data:")
    print(cna_data.index[:10].tolist())
    
    # Calculate signature scores
    print("\nCalculating signature scores...")
    signature_score = calc_signatures(edata, 
                                    os.path.join(data_path, "gene_signatures_20170111.gmt"),
                                    method="median")
    
    print("\nFirst 10 row labels in signature_score:")
    print(signature_score.index[:10].tolist())
    
    # Calculate CD103_Ratio
    print("Calculating CD103_Ratio...")
    CD103_pos = signature_score.loc["CD103_Positive_Median_Cancer.Cell.2014_PMID.25446897"]
    CD103_neg = signature_score.loc["CD103_Negative_Median_Cancer.Cell.2014_PMID.25446897"]
    CD103_ratio = CD103_pos - CD103_neg
    signature_score.loc["CD103_Ratio_Cancer.Cell.2014_PMID.25446897"] = CD103_ratio
    
    # Calculate differentiation score
    print("Calculating differentiation score...")
    diff_centroid = pd.read_csv(os.path.join(data_path, "special_gene_signature_training_sets/UNC_Differentiation.Score_Model_BCR.2010_PMID.20813035.txt"),
                               sep='\t', index_col=0)
    diff_score = assign_diff_score_dwd(diff_centroid, edata)
    signature_score.loc["UNC_Differentiation.Score_Model_BCR.2010_PMID.20813035"] = diff_score
    
    # Calculate Oncotype DX score
    print("Calculating Oncotype DX score...")
    oncotype = GHI_RS(edata)
    signature_score.loc["GHI_RS_Model_NJEM.2004_PMID.15591335"] = oncotype
    
    # Create results directory if it doesn't exist
    os.makedirs("results", exist_ok=True)
    
    # Save signature scores
    print("Saving signature scores...")
    signature_score.to_pickle("results/signature_score.pkl")
    
    # Calculate segment scores
    print("\nCalculating segment scores...")
    segment_score = calc_segments(cna_data, os.path.join(data_path, 'CNA_segments.utf8.gmt'), method='mean')
    
    # Save segment scores
    print("Saving segment scores...")
    segment_score.to_pickle("results/segment_score.pkl")
    
    print("\nAll calculations completed successfully!")

if __name__ == "__main__":
    main() 

File: APOLLO/DNADX/Reproduce/DNA-based-predictors-of-non-genetic-cancer-phenotypes/Rscript/signature_score_and_segment_score_calculation.R
# This script is for calculating gene signature scores based on RNA and segment scores based on gene-level DNA CNA

# Install and load required packages
if (!require("readxl")) {
    install.packages("readxl", repos="http://cran.r-project.org")
    library(readxl)
}
if (!require("GSA")) {
    install.packages("GSA", repos="http://cran.r-project.org")
    library(GSA)
}

# Source helper functions
source("Rscript/helper.R")

# Load input data
edata <- read.table("HiSeqV2", header=TRUE, sep="\t", check.names=FALSE)
CNdata <- read.table("Gistic2_CopyNumber_Gistic2_all_data_by_genes.txt", header=TRUE, sep="\t", check.names=FALSE)

# Convert data frames to matrices with proper row names
edata_matrix <- as.matrix(edata[,-1])  # Remove first column (gene names)
rownames(edata_matrix) <- as.character(edata[[1]])  # Use first column as row names

CNdata_matrix <- as.matrix(CNdata[,-1])  # Remove first column (gene names)
rownames(CNdata_matrix) <- as.character(CNdata[[1]])  # Use first column as row names

# --- Debug: Check GMT loading and initial signature calculation ---
cat("Attempting to load GMT file: data/gene_signatures_20170111.gmt\n")
temp_geneset_obj <- tryCatch(GSA.read.gmt("data/gene_signatures_20170111.gmt"), error = function(e) { cat("Error loading GMT file:", e$message, "\n"); NULL })

if (!is.null(temp_geneset_obj)) {
    cat("GMT file loaded. Number of gene sets found:", length(temp_geneset_obj$genesets), "\n")
    cat("First 10 gene set names from GMT:
")
    print(head(temp_geneset_obj$geneset.names, 10))
    # Check if your specific signature is in the loaded names
    target_gsea_sig <- "GSEA_Median_GP17_Basal_signaling_r0_958_SMID_BREAST_CANCER_BASAL_UP"
    if (target_gsea_sig %in% temp_geneset_obj$geneset.names) {
        cat(paste("Target signature '", target_gsea_sig, "' FOUND in GMT loaded names.\n"))
    } else {
        cat(paste("Target signature '", target_gsea_sig, "' NOT FOUND in GMT loaded names. Checking for similar names...\n"))
        # Very simple partial match check (case-insensitive)
        matches <- temp_geneset_obj$geneset.names[grepl("BASAL_SIGNALING", toupper(temp_geneset_obj$geneset.names))]
        if(length(matches) > 0) {
            cat("Potential matches found in GMT for 'BASAL_SIGNALING':\n")
            print(matches)
        }
    }
} else {
    cat("Failed to load or parse GMT file. calc_signatures will likely fail or produce empty results for GMT signatures.\n")
}
# -------------------------------------------------------------------

# Calculate signature scores
cat("Calculating signature scores from GMT...\n")
signature_score <- calc_signatures(edata_matrix, "data/gene_signatures_20170111.gmt", method = "median")

# --- Debug: Check signature_score immediately after calc_signatures ---
cat("\n--- signature_score immediately after calc_signatures (from GMT) ---\n")
if (exists("signature_score") && !is.null(signature_score) && !is.null(dim(signature_score))) {
    cat(paste("Dimensions:", paste(dim(signature_score), collapse=" x "), "\n"))
    cat("First 10 rownames (signatures) from GMT processing:
")
    print(head(rownames(signature_score), 10))
    # Check for your specific signature again
    if (target_gsea_sig %in% rownames(signature_score)) {
        cat(paste("Target signature '", target_gsea_sig, "' FOUND in initial signature_score object.\n"))
    } else {
        cat(paste("Target signature '", target_gsea_sig, "' NOT FOUND in initial signature_score object.\n"))
    }
} else {
    cat("signature_score object is NULL, empty, or not a matrix/data.frame after calc_signatures.\n")
}
cat("---------------------------------------------------------------------\n\n")
# ---------------------------------------------------------------------

# find NA signatures
NAsig <- rownames(signature_score[is.na(signature_score[,1]),])
# remove NA signatures
i <- match(NAsig,rownames(signature_score))
if (length(i) > 0) {
    signature_score <- signature_score[-i,]
}

# CD103_Ratio
CD103_pos <- signature_score[rownames(signature_score)=="CD103_Positive_Median_Cancer.Cell.2014_PMID.25446897"]
CD103_neg <- signature_score[rownames(signature_score)=="CD103_Negative_Median_Cancer.Cell.2014_PMID.25446897"]
CD103_ratio <- CD103_pos - CD103_neg # log2 scale division 
signature_score <- rbind(signature_score,CD103_ratio)
rownames(signature_score)[nrow(signature_score)] <- "CD103_Ratio_Cancer.Cell.2014_PMID.25446897"

# differentiation score
print("Reading differentiation score training data...")
diff_centroid <- read.table("/ix1/alee/LO_LAB/Personal/Daisong/Projects/APOLLO/DNADX/Reproduce/DNA-based-predictors-of-non-genetic-cancer-phenotypes/data/special_gene_signature_training_sets/UNC_Differentiation.Score_Model_BCR.2010_PMID.20813035.txt", sep = '\t', header = T, row.names = 1, check.names = F)
print("Structure of diff_centroid:")
print(str(diff_centroid))
print("Head of diff_centroid:")
print(head(diff_centroid))
print("Structure of edata_matrix:")
print(str(edata_matrix))
print("Head of edata_matrix:")
print(head(edata_matrix))

diff_score <- assignDiffScore.dwd(diff_centroid,edata_matrix)
signature_score <- rbind(signature_score,diff_score)
rownames(signature_score)[nrow(signature_score)] <- "UNC_Differentiation.Score_Model_BCR.2010_PMID.20813035"

# oncotype DX score
oncotype <- GHI_RS(edata_matrix)
signature_score <- rbind(signature_score,oncotype)
rownames(signature_score)[nrow(signature_score)] <- "GHI_RS_Model_NJEM.2004_PMID.15591335"

# Save results in the results directory
save(signature_score,file = 'results/signature_score.rda')

# Calculate segment scores
segment_score <- calc_segments(CNdata_matrix,'data/CNA_segments.gmt',method = 'mean')

# Save segment scores in the results directory
save(segment_score,file = 'results/segment_score.rda')





